File: 05/1.js
const strs = require("fs").readFileSync("input.txt").toString().split(",")
const INPUT = 1

for (let ip = 0; ip < strs.length;) {
    const [i, a1, a2, a3] = strs.slice(ip)
    const [p1 = "0", p2 = "0"] = i.split("").reverse().slice(2)
    
    if (i == "99") {
        break
    } else if (i.endsWith("4")) {
        console.log(+p1 ? a1 : strs[+a1])
        ip += 2
    } else if (i === "3") {
        strs[+a1] = INPUT
        ip += 2
    } else if (i.endsWith("1")) {
        strs[+a3] = String((+p1 ? +a1 : +strs[a1]) + (+p2 ? +a2 : +strs[a2]))
        ip += 4
    } else if (i.endsWith("2")) {
        strs[+a3] = String ((+p1 ? +a1 : +strs[a1]) * (+p2 ? +a2 : +strs[a2]))
        ip += 4
    } else {
        throw new Error(`Invalid OP Code '${i}'`)
    }
}

File: 23/1.hs
import IntCode
import Data.Map (Map)
import Data.Map qualified as M
import Data.Set (Set)
import Data.Set qualified as S
import Data.Maybe (fromMaybe)
import Data.Foldable (foldl')
import Text.Printf (printf)
import Debug.Trace

type Message = (Integer, Integer, Integer)

main = do
    input <- readFile "input.txt"
    let code = read $ "[" ++ input ++ "]"
    let makeState = initialState code
    let (to, x, y) = go 0 (M.fromList [(i, makeState [fromIntegral i]) | i <- [0..l - 1]]) M.empty
    printf "Part 1: %d\n" y
    where
        l = 50
        go :: Int -> Map Int State -> Map Int [Message] -> Message
        go idx machines messages = -- traceShow (M.filter (not . null) messages) $
            let machine = machines M.! idx
                state' = runUntilInput machine
                newMessages = chunksOf 3 . reverse . output $ state'
                receivers = S.fromList [ fromIntegral to | [to, _, _] <- newMessages]
                messages' = foldl' (flip insertMessage) messages newMessages
                messagesForMachine = fromMaybe [] (messages' M.!? idx)
                state'' =
                    if null messagesForMachine
                        then push (-1) $ state'
                        else foldl' (flip push) state' (concatMap (tail . messageToList) messagesForMachine)
                messages'' = M.insert idx [] messages'
                machines' = M.insert idx state'' machines 
            in -- traceShow (idx, newMessages, M.filterWithKey (\k _ -> S.member k receivers) messages, M.filterWithKey (\k _ -> S.member k receivers) messages') $
                case messages M.!? 255 of
                    Just (x:_) -> x
                    _ -> go (succ idx `mod` l) machines' messages''

        insertMessage :: [Integer] -> Map Int [Message] -> Map Int [Message]
        insertMessage [to, x, y] messages =
            case messages M.!? (fromIntegral to) of
                Just xs -> M.insert (fromIntegral to) (xs ++ [(to, x,y)]) messages
                Nothing -> M.insert (fromIntegral to) [(to, x,y)] messages

        messageToList (to, x, y) = [to, x, y]

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n xs
    | length xs < n = error $ "Cannot divide list to chunks"
    | otherwise = take n xs : chunksOf n (drop n xs)

